api/src/utils/async-request-storage.ts

```diff
@@ -0,0 +1,93 @@
+import { AsyncLocalStorage } from 'async_hooks';
+import { NextFunction, Request, Response } from 'express';
+import { v4 as uuid } from 'uuid';
+
+const DEFAULT_REQUEST_STORE_VALUE = 'SYSTEM';
+
+// Request store keys
+export type RequestStoreKey = 'requestId' | 'username';
+
+// Map (Store) for request metadata
+// Note: Intentionally using type `string` for values to restrict the potential data types,
+// trying to keep this as lightweight as possible.
+export type RequestStore = Map<RequestStoreKey, string>;
+
+// Initialize the request storage to track request metadata
+export const AsyncRequestStorage = new AsyncLocalStorage<RequestStore>();
+
+/**
+ * Middleware to initialize the async request storage.
+ *
+ * Why? This allows us to store request-specific data (ie: request id, username) in a key-value store
+ * that is accessible to all downstream code for the current request.
+ *
+ * Note: Must be initialized BEFORE any other middleware that requires the request storage.
+ * Specifically, this should be the first middleware in the middleware stack.
+ * ie: Inside the openapi 'x-express-openapi-additional-middleware' array.
+ *
+ * @see https://nodejs.org/api/async_context.html#class-asynclocalstorage
+ *
+ * @param {Request} req
+ * @param {Response} _res
+ * @param {NextFunction} next
+ * @return {*} {void}
+ */
+export function initRequestStorage(req: Request, _res: Response, next: NextFunction) {
+  const requestStore: RequestStore = new Map();
+
+  // Generate the request id for the current request - unique for each request
+  requestStore.set('requestId', uuid());
+
+  // Set the username of the user who made the current request
+  requestStore.set('username', req.keycloak_token?.idir_username ?? req.keycloak_token?.bceid_username);
```

Comment:
in `keycloak-utils.ts` there is an existing `getUserIdentifier` function, which I believe does the same thing here.

Reply:
Fixed!